Adding Discovery API for EndpointSlice

robscott opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR adds the Discovery API group for EndpointSlices, as mentioned in the associated [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/20190603-EndpointSlice-API.md) and [Enhancement Issue](https://github.com/kubernetes/enhancements/issues/752).

**Special notes for your reviewer**:
Since this PR contains a lot of generated code, I've kept the code generation separate in a second commit.

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

- [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/20190603-EndpointSlice-API.md)
- [Enhancement Issue](https://github.com/kubernetes/enhancements/issues/752)

robscott committed:
Adding discovery/v1alpha1 API for EndpointSlices

freehan commented:
/ok-to-test

freehan commented on pkg/apis/discovery/v1alpha1/defaults.go:
```diff
@@ -0,0 +1,43 @@
+}
+
+func SetDefaults_EndpointSlice(obj *discoveryv1alpha1.EndpointSlice) {
+	// No defaults at this level yet
```
Need default for TargetType.

robscott replied:
Good catch, thanks! Added default + validation along with related tests.

BenTheElder commented:
/uncc
I don't think i'm appropriate to review this, happy to help approve later after review if need be though!

freehan commented:
/label api-review

robscott commented:
/retest

thockin commented on pkg/apis/discovery/validation/validation.go:
```diff
@@ -0,0 +1,86 @@
+)
+
+func ValidateEndpointSlice(endpointSlice *discovery.EndpointSlice) field.ErrorList {
+	allErrs := apivalidation.ValidateObjectMeta(&endpointSlice.ObjectMeta, true, apivalidation.ValidateReplicationControllerName, field.NewPath("metadata"))
```
define your own name validator:

```
var validateEndpointSliceName = apimachineryvalidation.NameIsDNSSubdomain

thockin commented on pkg/apis/discovery/validation/validation.go:
```diff
@@ -0,0 +1,86 @@
+
+	if endpointSlice.TargetType == nil {
+		allErrs = append(allErrs, field.Required(field.NewPath("targetType"), ""))
+	}
```
} else if *endpointSlice.TargetType != discovery.IPTargetType {

thockin commented on pkg/apis/discovery/validation/validation.go:
```diff
@@ -0,0 +1,86 @@
+	}
+
+	if endpointSlice.TargetType != nil && *endpointSlice.TargetType != discovery.IPTargetType {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("targetType"), *endpointSlice.TargetType, "Must be IP"))
```
follow examples for protocol:

```
var supportedPortProtocols = sets.NewString(string(core.ProtocolTCP), string(core.ProtocolUDP), string(core.ProtocolSCTP)) 
```

```
    } else if !supportedPortProtocols.Has(string(port.Protocol)) {
        allErrs = append(allErrs, field.NotSupported(fldPath.Child("protocol"), port.Protocol, supportedPortProtocols.List()))                    
```

Specificall "not supported"

thockin replied:
error messages start lower-case

thockin commented on pkg/apis/discovery/validation/validation.go:
```diff
@@ -0,0 +1,86 @@
+		idxPath := fldPath.Index(i)
+
+		if len(endpoint.Targets) < 1 {
+			allErrs = append(allErrs, field.Required(idxPath.Child("targets"), "At least 1 target required"))
```
messages start lower-case

thockin commented on pkg/apis/discovery/validation/validation.go:
```diff
@@ -0,0 +1,86 @@
+
+	for i, endpointPort := range *endpointPorts {
+		idxPath := fldPath.Index(i)
+
```
first check if nil - if not, field.Required()

thockin commented on pkg/apis/discovery/validation/validation.go:
```diff
@@ -0,0 +1,86 @@
+			portNames.Insert(*endpointPort.Name)
+		}
+
+		if len(*endpointPort.Protocol) == 0 {
```
split logic: nil = field.Required(), "" == field.Invalid()

thockin commented on pkg/registry/discovery/endpointslice/strategy.go:
```diff
@@ -0,0 +1,93 @@
+	// Any changes to the endpoints increment the generation number, any changes to the
+	// status should reflect the generation number of the corresponding object.
+	// See metav1.ObjectMeta description for more information on Generation.
+	if !apiequality.Semantic.DeepEqual(oldEndpointSlice.Endpoints, newEndpointSlice.Endpoints) {
```
why just Endpoints - not ports?

thockin commented on pkg/registry/discovery/endpointslice/strategy.go:
```diff
@@ -0,0 +1,93 @@
+
+// ValidateUpdate is the default update validation for an end user.
+func (endpointSliceStrategy) ValidateUpdate(ctx context.Context, obj, old runtime.Object) field.ErrorList {
+	return validation.ValidateEndpointSlice(obj.(*discovery.EndpointSlice))
```
This probably needs a separate validation.  Even if the logic ends up being the same, make a ValidateEndpointSliceUpdate() func and call the other logic, so it is obvious.

For example, is "TargetType" mutable?  If so, you need to verify that all targets are the correct type.

thockin replied:
Are there any other things that could go wrong when mutating in-place?  Think hard about this and try to break it.

robscott replied:
Good call, will separate this out. We've been considering all EndpointSlice specific fields (excluding meta) mutable. The EndpointSlice controller relies on this when it "recycles" an EndpointSlice marked for deletion to be reused in place of one marked for creation. This reduces create and delete actions to simple update ones.

thockin replied:
the only obvious one I saw was that `targetType` has to be validated against all the `target` values (or vice-versa, really).  It's probably OK to make them mutable, as long as they are co-valid

robscott replied:
Yeah that makes sense. I've added validation to ensure `addressType` is immutable on updates.

thockin replied:
That will probably not impact the controller, since it will only be re-using across a single type anyway, right?  I'm fine with this being mutable, as long as the validation catches it (and we have to write the validation ANYWAY, so...)

Your call

thockin commented on pkg/apis/discovery/validation/validation.go:
```diff
@@ -0,0 +1,86 @@
+		if len(endpoint.Targets) < 1 {
+			allErrs = append(allErrs, field.Required(idxPath.Child("targets"), "At least 1 target required"))
+		}
+	}
```
validate every target against targetType - parse it as an IP and check for nil

thockin commented on staging/src/k8s.io/api/discovery/v1alpha1/types.go:
```diff
@@ -0,0 +1,125 @@
+limitations under the License.
+*/
+
+package v1alpha1
```
all same comments from internal

thockin reviewed (commented):
I have a few notes.

In general we don't want to merge API until the implementation is done.  For an alpha maybe, but we don't want to release a system with a half-cooked API, even alpha really.

robscott commented on pkg/apis/discovery/types.go:
```diff
@@ -0,0 +1,124 @@
+	// Different consumers (e.g. kube-proxy) handle different types of
+	// targets in the context of its own capabilities.
+	// +listType=set
+	Targets []string
```
I like "address" better - updated to use that everywhere including addressType in place of targetType.

freehan commented on pkg/apis/discovery/types.go:
```diff
@@ -0,0 +1,124 @@
+	// Standard object's metadata.
+	// +optional
+	metav1.ObjectMeta
+	// This field specifies the list of ports associated with each endpoint in the EndpointSlice
```
Can we do this?

`Ports = nil` means "port does not apply". Can be applied to AddressType that does not have port.

`Ports = []` means "no port specified". This happens when there is no pods behind a service.

thockin replied:
No, please do not try to do that.  Not every transport or language can distinguish those (e.g. proto requires jumping through an extra message)

freehan commented on pkg/apis/discovery/types.go:
```diff
@@ -0,0 +1,124 @@
+// EndpointConditions represents the current condition of an endpoint
+type EndpointConditions struct {
+	// Ready indicates if the endpoint is ready to serve traffic
+	Ready bool
```
From what I can tell, `don't know` has a tendency to cover lots of different states. (e.g. endpoint has not been health checked vs. endpoint has not been programmed) Hence, making it hard to consume. The consumer is likely to fall back to ready/notReady

thockin replied:
What I am afraid of is that there could be cases where the caller legit does not know, and we're forcing them to say yes.  Alternatively we could say nil means "don't know" and callers should treat it as a tentative yes

thockin replied:
or not - not super locked into this one

thockin replied:
it just felt like a field I shouldn't have to specify if I don't know.

freehan replied:
You are right. Maybe making the Ready field an optional pointer?

thockin replied:
I'm truly torn on this.  If we make it optional, what will kube-proxy do in case of `nil`?  Maybe it's not such a hardship to require users to set this, given that most users are controllers...  Or we could default it to true, so only "clever" controllers which add not-yet-ready addresses need to do anything.

What do you think?  You guys have as much voice in this as I do.

freehan replied:
Just a recap: A producer of endpointSlice can leave a condition nil if it "does-not-apply" or "does-not-know". Then the condition meant "unknown" or "unspecified" to the consumer. Then it is up to the consumer to interpret the condition (the consumer may not care about the condition). I feel like we can only enforce kube-proxy/kube-dns to interpret "not specified" as ready. 

That sort of mean we can only document, "consumer **should** interpreted as yes" instead of "must".  WDYT?

thockin replied:
Yeah, that's pretty much my (B), but more accurate.

thockin commented on pkg/registry/discovery/endpointslice/strategy.go:
```diff
@@ -0,0 +1,99 @@
+
+// PrepareForUpdate clears fields that are not allowed to be set by end users on update.
+func (endpointSliceStrategy) PrepareForUpdate(ctx context.Context, obj, old runtime.Object) {
+	newEndpointSlice := obj.(*discovery.EndpointSlice)
```
FWIW go conventions prefer shorter names for local vars.  oldEPS newEPS would be OK :)

thockin commented on pkg/registry/discovery/endpointslice/strategy.go:
```diff
@@ -0,0 +1,99 @@
+	// number, any changes to the status should reflect the generation number of
+	// the corresponding object. See metav1.ObjectMeta description for more
+	// information on Generation.
+	incrementGeneration := apiequality.Semantic.DeepEqual(oldEndpointSlice.Endpoints, newEndpointSlice.Endpoints)
```
would this be simpler as:

```
    newEndpointSlice := obj.(*discovery.EndpointSlice)
￼	oldEndpointSlice := old.(*discovery.EndpointSlice)

    newEndpointSlice.ObjectMeta = v1.ObjectMeta{}  // whiteout
    incrementGeneration := apiequality.Semantic.DeepEqual(oldEndpointSlice, newEndpointSlice)

thockin commented on pkg/apis/discovery/types.go:
```diff
@@ -0,0 +1,140 @@
+	metav1.ObjectMeta
+	// endpoints is a list of unique endpoint objects in this slice.
+	// +listType=atomic
+	Endpoints []Endpoint
```
Do we want to specify a limit to number of Endpoints in a slice?  We never get another chance.  1000?  2000?  5000?

@freehan

freehan replied:
It will hit the object storage limit naturally. Which means it is around 5000 in normal cases. 

So we might as well set it to 1000?

thockin replied:
1000 seems reasonable.  We can lift limits if we need to.  Let's define it and validate it.  We have `field.TooLong` but that's not right.  Maybe we want to add `field.TooMany()` ?

robscott replied:
I added field.TooMany() in my latest commit, hopefully it makes sense.

robscott replied:
(1000 is also set as the limit here now)

thockin commented on pkg/apis/discovery/types.go:
```diff
@@ -0,0 +1,140 @@
+	// when no pods match a Service selector and a placeholder EndpointSlice is
+	// generated.
+	// +optional
+	// +listType=atomic
```
do we want to specify a limit to number of ports?  100?  200?  500?

freehan replied:
This needs to match services though. Without port range, I would imagine many people specifying many ports.

thockin replied:
There's a practical limit that we want to defend, and if it overflows it could just become a new slice, right?  E.g. if we limit to 100 ports and I need 101, make 2 slices? Even if we start at 500, we'll need some overflow handling... (and tests)

robscott replied:
I added an initial limit of 100 here, easy to change the value if another number makes more sense.

thockin commented on pkg/apis/discovery/types.go:
```diff
@@ -0,0 +1,140 @@
+	// addresses. For example, a dual-stack system might have both IPv4 and IPv6
+	// addresses.
+	// +listType=set
+	Addresses []string
```
do we want to specify a limit to number of addresses?  100?  200?  500?

freehan replied:
Can we say AddressType==IP, then at most 2 Address?

And then we extend the validation based on AddressType

thockin replied:
No, there are people doing multi-IP stuff.  In fact, that may be a reason to change this to 

```
Addresses []EndpointAddress
```
```
type EndpointAddress struct {
    address strring
}
```

So we can later copy metadata from Pod.IPs[].

Thoughts?  It's kind of obnoxious and would add another dimension of inefficiency.  I don't know if we'll need it so we can revisit as we progress alpha.

How about a limit of 100 with validation?

thockin replied:
For IPs, 16 would probably be enough.  4 would probably be enough, but let's just set it big enough that nobody complains?

robscott replied:
I added an initial limit of 100 here.

robscott replied:
I added an additional limit of 100 here, happy to change to something else if it makes more sense

thockin reviewed (commented):
2nd pass:

I spent most of my time making sure comments are clear and complete and that corner-cases are documented and sane.  Also on validation.

Run through this as if you were trying to be as antagonistic as possible.  What else is under-specified?  What could a user possibly do to break things?  Without limits, what unexpected scenario will cause problems later.  Remember we can always lift restrictions, but we can't impose them.

freehan commented on pkg/apis/discovery/types.go:
```diff
@@ -0,0 +1,140 @@
+	// ports specifies the list of network ports exposed each endpoint in this
+	// EndpointSlice. Each port must have a unique name. When endpoints is nil,
+	// it indicates that ports are irrelevant for the address type. When this is
+	// an empty slice, it indicates that ports are relevant for the address type,
```
Okay. Sounds good.

freehan commented on pkg/apis/discovery/types.go:
```diff
@@ -0,0 +1,140 @@
+	// generated.
+	// +optional
+	// +listType=atomic
+	Ports []EndpointPort
```
I feel like the port number could not be unique right? Since we allow different service port map to the same target port.

svcPort1 -> 80
svcPort2 -> 80

thockin replied:
good point.

robscott commented on pkg/apis/discovery/types.go:
```diff
@@ -0,0 +1,140 @@
+	// The name of this port. All ports must have a unique name.
+	// If the EndpointSlice is dervied from K8s service, this corresponds to
+	// the Service.ports[].name from a Kubernetes Service.
+	// Name must be a IANA_SVC_NAME or an empty string.
```
I may have gone a bit too far in the other direction now, wasn't sure if I should list the validation criteria or just RFC here.

thockin replied:
this looks good

robscott commented on pkg/apis/discovery/v1alpha1/defaults.go:
```diff
@@ -0,0 +1,47 @@
+
+func SetDefaults_EndpointSlice(obj *discoveryv1alpha1.EndpointSlice) {
+	if obj.AddressType == nil {
+		ipAddressType := discoveryv1alpha1.AddressTypeIP
```
much better, thanks!

robscott commented on pkg/apis/discovery/validation/validation.go:
```diff
@@ -0,0 +1,124 @@
+	for i, endpointPort := range *endpointPorts {
+		idxPath := fldPath.Index(i)
+
+		// Certain code paths could reach here without defaults set here ("").
```
Good to know that it's safe to assume resources will only reach here after defaulting in normal use cases. The potential nil pointer felt weird to leave unchecked, but my solution was less than ideal, thanks for clarifying!

robscott commented on pkg/apis/discovery/validation/validation.go:
```diff
@@ -0,0 +1,124 @@
+
+		if endpointPort.Protocol == nil {
+			allErrs = append(allErrs, field.Required(idxPath.Child("protocol"), ""))
+		} else if endpointPort.Protocol == nil {
```
haha not quite sure how I did that :)

thockin commented on pkg/apis/discovery/types.go:
```diff
@@ -0,0 +1,140 @@
+	// +listType=atomic
+	Endpoints []Endpoint
+	// ports specifies the list of network ports exposed by each endpoint in
+	// this slice. Each port must have a unique name. When ports is empty, it
```
now it says the unique-names requirement twice, LOL.  Sorry.

robscott replied:
lol oh my, that should finally be fixed, I hope.

thockin commented on pkg/apis/discovery/types.go:
```diff
@@ -0,0 +1,140 @@
+	// * may contain only [-a-z0-9]
+	// * must contain at least one letter [a-z]
+	// * it must not start or end with a hyphen, nor contain adjacent hyphens
+	// +optional
```
as discussed, we should not denote this as optional

thockin commented on pkg/registry/discovery/endpointslice/strategy.go:
```diff
@@ -0,0 +1,96 @@
+	newEPS := obj.(*discovery.EndpointSlice)
+	oldEPS := old.(*discovery.EndpointSlice)
+
+	// Increment generation if endpoints, ports, or address type have changed
```
If you want to open-code the list of fields, add a comment that this needs to be updated if any fields are added to EndpointSlice

thockin replied:
That's why I suggested a white-out approach instead

robscott replied:
I did try a couple variations of that but kept on getting failing tests since the meta did not seem to be sticking, despite my efforts to reassign it after the whiteout. It didn't make any sense, will definitely revisit, just wanted to get tests back to passing for now.

thockin replied:
@liggitt @lavalamp Who has best context on use of `generation`?  Why don't we do it automatically?  Is there a safe best-practice other than open white-listing fields (likely to fail if new fields are added)?

liggitt replied:
most types that use generation have a spec/status split, and do a deepequal comparison on the entire spec to decide whether to bump generation:

```
	if !apiequality.Semantic.DeepEqual(oldDaemonSet.Spec, newDaemonSet.Spec) {
		newDaemonSet.Generation = oldDaemonSet.Generation + 1
	}
```

```
	if !apiequality.Semantic.DeepEqual(oldRS.Spec, newRS.Spec) {
		newRS.Generation = oldRS.Generation + 1
	}
```


> Why don't we do it automatically?

things like annotations/labels/status generally shouldn't bump generation (especially status, otherwise status.observedGeneration would never be able to match generation)

robscott replied:
Ended up getting the white-out approach working, we'll just have to change/update that if we ever add something like a status field that we don't want to increment generation for.

thockin reviewed (commented):
3rd pass

Less this time!

thockin commented:
So I think the right answer here is to figure out why a white-out approach
doesn't work and fix that.

On Fri, Aug 16, 2019 at 8:54 AM Jordan Liggitt <notifications@github.com>
wrote:

> *@liggitt* commented on this pull request.
> ------------------------------
>
> In pkg/registry/discovery/endpointslice/strategy.go
> <https://github.com/kubernetes/kubernetes/pull/80766#discussion_r314782764>
> :
>
> > +func (endpointSliceStrategy) NamespaceScoped() bool {
> +	return true
> +}
> +
> +// PrepareForCreate clears the status of an EndpointSlice before creation.
> +func (endpointSliceStrategy) PrepareForCreate(ctx context.Context, obj runtime.Object) {
> +	endpointSlice := obj.(*discovery.EndpointSlice)
> +	endpointSlice.Generation = 1
> +}
> +
> +// PrepareForUpdate clears fields that are not allowed to be set by end users on update.
> +func (endpointSliceStrategy) PrepareForUpdate(ctx context.Context, obj, old runtime.Object) {
> +	newEPS := obj.(*discovery.EndpointSlice)
> +	oldEPS := old.(*discovery.EndpointSlice)
> +
> +	// Increment generation if endpoints, ports, or address type have changed
>
> most types that use generation have a spec/status split, and do a
> deepequal comparison on the entire spec to decide whether to bump
> generation:
>
> 	if !apiequality.Semantic.DeepEqual(oldDaemonSet.Spec, newDaemonSet.Spec) {
> 		newDaemonSet.Generation = oldDaemonSet.Generation + 1
> 	}
>
> 	if !apiequality.Semantic.DeepEqual(oldRS.Spec, newRS.Spec) {
> 		newRS.Generation = oldRS.Generation + 1
> 	}
>
> Why don't we do it automatically?
>
> things like annotations/labels/status generally shouldn't bump generation
> (especially status, otherwise status.observedGeneration would never be able
> to match generation)
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/80766?email_source=notifications&email_token=ABKWAVCIIGV3NY2XCVJCBZDQE3EUTA5CNFSM4IIARHY2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCBZ6MZY#discussion_r314782764>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABKWAVE6YEPCSX5OPQUVOF3QE3EUTANCNFSM4IIARHYQ>
> .
>

robscott commented on pkg/apis/discovery/validation/validation.go:
```diff
@@ -0,0 +1,120 @@
+		addrType = *endpointSlice.AddressType
+	}
+
+	allErrs = append(allErrs, validateEndpoints(&endpointSlice.Endpoints, addrType, field.NewPath("endpoints"))...)
```
good call, thanks!

robscott commented on pkg/apis/discovery/types.go:
```diff
@@ -0,0 +1,140 @@
+	// endpoint.
+	// Key/value pairs contained in topology must conform with the label format.
+	// https://kubernetes.io/docs/concepts/overview/working-with-objects/labels
+	// Topology may include, but is not limited to the following well known keys:
```
Added validation for a max of 16 + standard label validation

thockin commented on staging/src/k8s.io/api/discovery/v1alpha1/types.go:
```diff
@@ -0,0 +1,141 @@
+	Endpoints []Endpoint `json:"endpoints" protobuf:"bytes,2,rep,name=endpoints"`
+	// ports specifies the list of network ports exposed by each endpoint in
+	// this slice. Each port must have a unique name. When ports is empty, it
+	// indicates that there are no defined ports. The most common example of
```
Is "The most common example" needed here?  It seems to be emphasizing a relatively corner-case and exposing impl details about one potential consumer of this API.  We should instead just say what we expect consumers to do if the find >0 endpoints and 0 ports.  E.g. "..no defined ports.  Consumers should interpret this as 'all ports'."

This fails-open, but again with POLS I think this is probably right.  If there's a security concern with publishing ports (security through obscurity) you have a different problem.  How does that fit?

freehan replied:
> what we expect consumers to do if the find >0 endpoints and 0 ports. E.g. "..no defined ports. Consumers should interpret this as 'all ports'."

Can we do this? 

- **">0 enpdoints and 0 ports"**: Up to the consumer to interpret. (Ports may be irrelevant)
- **">0 endpionts and 1 EndpointPort with nil port"**: Consumer must interpret as "all ports"

thockin replied:
That seems fine.  We can smooth out exact wording once it is in.

Specifically, no ports[] means both protocols and ports are not defined - it's an address with no other information.  ports[] with no port number means "all ports for the specified protocol", because protocol will be defaulted.

The point here is not to leave the meaning of these cases up for debate.

thockin commented on staging/src/k8s.io/api/discovery/v1alpha1/types.go:
```diff
@@ -0,0 +1,141 @@
+	// All addresses in this slice must be the same type.
+	// Default is IP
+	// +optional
+	AddressType *AddressType `json:"addressType" protobuf:"bytes,4,rep,name=addressType"`
```
nit: for readability, I would put this about endpoints (and move the definition of it just below this struct)

thockin commented on staging/src/k8s.io/api/discovery/v1alpha1/types.go:
```diff
@@ -0,0 +1,141 @@
+	Items []EndpointSlice `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// AddressType represents the type of address referred to by an Endpoint.
```
don't capitalize Endpoint

thockin commented on staging/src/k8s.io/api/discovery/v1alpha1/types.go:
```diff
@@ -0,0 +1,141 @@
+	// according to the corresponding EndpointSlice addressType field.
+	// Different consumers (e.g. kube-proxy) handle different types of addresses
+	// in the context of their own capabilities. This must contain at least one
+	// address but no more than 100.
```
Let's add something like:

```
...according to the corresponding EndpointSlice addressType field.  This allows for 
cases like dual-stack (IPv4 and IPv6) networking.  Consumers (e.g. kube-proxy) 
must handle different types of addresses ...
```

thockin commented on pkg/apis/discovery/validation/validation.go:
```diff
@@ -0,0 +1,138 @@
+	allErrs := field.ErrorList{}
+
+	if len(endpoints) > maxEndpoints {
+		allErrs = append(allErrs, field.TooMany(fldPath, endpoints, maxEndpoints))
```
Maybe we should terminate here, rather than process a list we know is too long - what think?  If so, same in other places.

thockin commented on pkg/apis/discovery/validation/validation.go:
```diff
@@ -0,0 +1,138 @@
+		addressPath := idxPath.Child("addresses")
+
+		if addrType == discovery.AddressTypeIP {
+			if len(endpoint.Addresses) < 1 {
```
nit: `== 0` is easier to read

thockin commented on staging/src/k8s.io/apimachinery/pkg/util/validation/field/errors.go:
```diff
@@ -201,6 +207,13 @@ func TooLong(field *Path, value interface{}, maxLength int) *Error {
+// TooMany returns a *Error indicating "too many". This is used to
+// report that a given list has too many items. This is similar to TooLong,
+// but the returned error indicates quantity instead of length.
+func TooMany(field *Path, value interface{}, maxQuantity int) *Error {
```
We can't put value in the error - it's too big!  Instead pass the len() and store that.

thockin commented on pkg/apis/discovery/validation/validation_test.go:
```diff
@@ -0,0 +1,329 @@
+				ObjectMeta:  standardMeta,
+				AddressType: &addressTypeIP,
+				Ports: []discovery.EndpointPort{{
+					Name:     &nameEmpty,
```
IMO, this test would be more readable with some helper functions, e.g. `newString(v string) { return &v }` or even `makePort(name string, proto string, num int) { return discovery.EndpointPort{ ... } }`

This is especially true for the max-len tests.  `makeSSMap(16) map[string]string` and then internally iterate and produce keys like "x1", "x2", x3"...

thockin commented on pkg/apis/discovery/validation/validation_test.go:
```diff
@@ -0,0 +1,329 @@
+		isExpectedFailure bool
+		endpointSlice     *discovery.EndpointSlice
+	}{
+		"good-slice": {
```
missing cases with multiple ports and multiple endpoints and multiple addresses.

All should probably have cases for 0, 1, and max, plus failure cases for max+1

thockin reviewed (commented):
Almost there.  Picked at testing some.  There's already a lot of LOC there, but much of it is redundant boilerplate and missing a bunch of cases, I think.

thockin reviewed (commented):
API LGTM. 

How do we want to proceed with implementation?  Do you want to squash this down to 1 commit and then layer those PRs on top of it?

robscott commented:
@thockin Thanks! I've squashed this PR down to 1 commit and rebased the related PRs (https://github.com/kubernetes/kubernetes/pull/81048 and https://github.com/kubernetes/kubernetes/pull/81430) on that.

thockin commented:
/lgtm

/hold

tedyu commented on pkg/apis/discovery/types.go:
```diff
@@ -0,0 +1,139 @@
+}
+
+// EndpointConditions represents the current condition of an endpoint.
+type EndpointConditions struct {
```
Since one endpoint is represented, should the trailing 's' be removed ?
